### ISSUE TITLE
node-sass + libsass source coupling

### ISSUE THREAD

**Issue by saper** [sass/node-sass#744]
For better packaging I would like to have a complete source of the project in a single zip file.
Normally this is not a problem in Github, since I can simply pulls the zip file of every commit:
(e.g. https://github.com/sass/node-sass/archive/v2.0.1.zip)

Due to our use of submodules, however, the zipfile does not include a proper version of `libsass` as well as test specs are missing (I normally do not need them).

There are few possible solutions:

1. include `libsass` in our git and have one .zip file including libsass and our node bindings.

2. download two zipfiles separately, e.g. https://github.com/sass/node-sass/archive/v2.0.1.zip and https://github.com/sass/libsass/archive/3.1.0.zip

3. teach `node-sass` to use an out-of-tree libsass (use extracted sources or use a library)

My preferred solution would be 3, is that feasible?

**Comment by am11** [comment]
Thanks for your proposal.
This is similar to https://github.com/sass/node-sass/issues/389, which is actively under consideration at https://github.com/sass/node-sass/issues/712 and ultimately tackled by https://github.com/sass/node-sass/pull/743. Lets discuss it at #712.


The idea really is to make user point to any libsass commit and recompile the binary. The steps which our script might follow would be:

```bash
# in node-sass root directory
#
# when user will run `node scripts/update`
#
# read the value of npmSassConfig.libsass
# from package.json, which is the SHA1 commit hash
# to libsass. Then:
git clone https://github.com/sass/libsass
git reset --hard my_favourite_SHA1_commit_hash
npm update    # installs missing dev dependencies
node scripts/build -f
```

I think people can do all that today manually. This little automation will not bringing an extraordinary advantage anyway..


---

> Due to our use of submodules, however, the zipfile does not include a proper version of  libsass 

What do you mean, can you expand on it? The packaged libsass code is always pointing to the version of libsass we are releasing node-sass against.

> as well as test specs are missing

Yes and new we have removed the entire `test` directory from the production package for vNext. User can always use `git` to pull it. The idea really is to keep the package size as slim as possible.

Now that you have mentioned, I will remove libsass from package as well. :wink:

**Comment by saper** [comment]
I read #712 and #743 and came to the conclusion they are about *binary* package downloads, i.e. package with an already compiled binding file. 

What most packagers need is just a version number and a source file corresponding to this version.

During an automated build it is much simpler to get a .zip file than play with git. I have an automated build that installs clean-room operating system, nodejs/iojs, npm and all dependencies - it fetches
source zipfiles, produces binary packages and installs them as necessary. This is all automated. I do not fetch python, node and other dependencies from git - the build system compiles them once into the binary package and unless something changes I am using those packages for testing. Each packages has a version number, not Git hash.

Besides, users of the package want to know that they have 2.0.1 and they can upgrade to, say, 2.0.2 and not use git hashes only. (Again, this is not a binary executable as in http://github.com/sass/node-sass-binaries but a complete binary package which can be installed with `tar`, without even calling `npm`).

This is similar to #389 problem, but instead of checking some code into `src/libsass` using git it might be better to point to existing libsass sources or even compiled binary. If we don't want to link shared libsass library, we can always use shared one (libsass.a).

**Comment by am11** [comment]
> What most packagers need is just a version number and a source file corresponding to this version.

Today with v2.0.1, if you try `require('node-sass').info()`, you get node-sass and the corresponding libsass versions. Also you can `require('path/to/node-sass/package.json').libsass` to get only libsass version. See [package.json#L4](https://github.com/sass/node-sass/blob/beffe92723c2a1848eef248b27c79dac2bcf6446/package.json#L4).

**Comment by saper** [comment]
But this works only after the the module has been built...

~~~~
> require('node-sass').info()
Error: `libsass` bindings not found. Try reinstalling `node-sass`?
    at getBinding (/home/saper/sw/test2/node_modules/node-sass/lib/index.js:22:11)
    at Object.<anonymous> (/home/saper/sw/test2/node_modules/node-sass/lib/index.js:188:23)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at repl:1:1
    at REPLServer.defaultEval (repl.js:132:27)
~~~~

I'll try two-zip approach and will see where I can arrive at.

**Comment by am11** [comment]
You can also try `require('./package.json').libsass`.

**Comment by kevinSuttle** [comment]
Walking into this error, brought me here. How do we fix this today, and avoid this in the future?
```
Error: `libsass` bindings not found. Try reinstalling `node-sass`?
```

**Comment by am11** [comment]
@kevinSuttle, sorry to tell you this but you landed on an unrelated issue. You can search the issue tracker and find the reasons of that exception.

**Comment by am11** [comment]
Closing it as a direct duplicate of https://github.com/sass/node-sass/issues/389.
@saper, lets continue this discussion there.
